### PR TITLE
feat(models): add configurable residual connections in enc-proc-dec

### PR DESCRIPTION
**Pull request by JPXKQX**
## Description

This is a follow-up to PR #451 . The branch has been moved to anemoi-core (so we can now run integration tests), and rebased with main.

- It abstracts the residual connection around a new class, `BaseResidualConnection`. It currently supports 3 options: `SkipConnection`, `TruncatedConnection`, and `NoConnection`.
- It also creates the `SparseProjector` class which handles the projection logic. This class could be used in the future for the multi-scale loss or other use cases.

## What problem does this change solve?

Encapsulation of residual connection functionality.

***As a contributor to the Anemoi framework, please ensure that your changes include unit tests, updates to any affected dependencies and documentation, and have been tested in a parallel setting  (i.e., with multiple GPUs). As a reviewer, you are also responsible for verifying these aspects and requesting changes if they are not adequately addressed. For guidelines about those please refer to https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the [Contributor License Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md)

----
📚 Documentation preview 📚: https://anemoi-training--670.org.readthedocs.build/en/670/

----
📚 Documentation preview 📚: https://anemoi-graphs--670.org.readthedocs.build/en/670/

----
📚 Documentation preview 📚: https://anemoi-models--670.org.readthedocs.build/en/670/

### CODE REVIEW

File: models/src/anemoi/models/layers/residual.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+from torch import nn
+from torch_geometric.data import HeteroData
+
+from anemoi.models.distributed.graph import gather_channels
+from anemoi.models.distributed.graph import shard_channels
+from anemoi.models.distributed.shapes import apply_shard_shapes
+from anemoi.models.layers.sparse_projector import build_sparse_projector
+
+
+class BaseResidualConnection(nn.Module):
```
Comment (HCookie) [inline models/src/anemoi/models/layers/residual.py]: This should subclass or metaclass from ABCMeta

File: models/src/anemoi/models/layers/residual.py
Comment (HCookie) [inline models/src/anemoi/models/layers/residual.py:1]: License?

File: models/src/anemoi/models/layers/residual.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    """Base class for residual connection modules."""
+
+    def __init__(self, *args, **kwargs):
```
Comment (HCookie) [inline models/src/anemoi/models/layers/residual.py]: What's the reason for these wild card ignores?

File: models/src/anemoi/models/layers/residual.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+class SkipConnection(BaseResidualConnection):
+    """Skip connection module
+
+    It selects the most recent timestep from the input sequence.
+
+    This module is used to bypass processing layers and directly pass the latest input forward.
+    """
+
+    def __init__(self, step: int = -1, *args, **kwargs):
```
Comment (HCookie) [inline models/src/anemoi/models/layers/residual.py]: Again, wild card ignores.
This can lead to hard to debug states.

File: models/src/anemoi/models/layers/residual.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+    def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
+        """Return a zero tensor with the same shape as the last timestep."""
+        return torch.zeros_like(x[:, 0, ...], device=x.device, dtype=x.dtype)
+
+
+class TruncatedConnection(nn.Module):
+    """Truncated skip connection
+
+    It applies a coarse-graining and reconstruction of input features using sparse projections to
```
Comment (HCookie) [inline models/src/anemoi/models/layers/residual.py]: Is 2nd person best here?
I would expect more like the form of `This applies'

File: models/src/anemoi/models/layers/residual.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+    This module returns a zero tensor with the same shape as the last timestep.
+    """
+
+    def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
+        """Return a zero tensor with the same shape as the last timestep."""
+        return torch.zeros_like(x[:, 0, ...], device=x.device, dtype=x.dtype)
+
+
+class TruncatedConnection(nn.Module):
```
Comment (HCookie) [inline models/src/anemoi/models/layers/residual.py]: If this is a residual, why does it not subclass from the base?

File: models/src/anemoi/models/layers/residual.py
Comment (HCookie) [inline models/src/anemoi/models/layers/residual.py:1]: Overall the interface feels a bit clunky with massive wild card ignores, if this is inline with other parts of models I can overlook it, but I'd like to see some more thought on the abstraction to prevent these wildcards

File: models/src/anemoi/models/layers/sparse_projector.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    norm = total.scatter_add_(0, edge_index[1].long(), weights)
+    norm = norm[edge_index[1]]
+    return weights / (norm + 1e-8)
+
+
+class SparseProjector(torch.nn.Module):
+    """Constructs and applies a sparse projection matrix for mapping features between grids.
+
+    The projection matrix is constructed from edge indices and edge attributes (e.g., distances),
+    with optional row normalization.
```
Comment (HCookie) [inline models/src/anemoi/models/layers/sparse_projector.py]:
```suggestion
    with optional row normalisation.
```
British english correction

File: models/src/anemoi/models/layers/sparse_projector.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        edges_name: str,
+        edge_weight_attribute: Optional[str] = None,
+        src_node_weight_attribute: Optional[str] = None,
+        **kwargs,
+    ) -> "SparseProjector":
+        """Build a SparseProjection from a graph.
+
+        Parameters
+        ----------
+        graph : networkx.Graph
```
Comment (HCookie) [inline models/src/anemoi/models/layers/sparse_projector.py]: Is this the correct type hint?

File: models/src/anemoi/models/layers/sparse_projector.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+            **kwargs,
+        )
+
+    def forward(self, x, *args, **kwargs):
+        # This has to be called in the forward because sparse tensors cannot be registered as buffers,
+        # as they can't be broadcast correctly when using DDP.
+        self.projection_matrix = self.projection_matrix.to(x.device)
+
+        out = []
+        device_type = "cuda" if torch.cuda.is_available() else "cpu"
```
Comment (HCookie) [inline models/src/anemoi/models/layers/sparse_projector.py]: This should use the type of the `x.device`?

File: models/src/anemoi/models/layers/residual.py (hunk 2 above)
Comment (JPXKQX) [reply models/src/anemoi/models/layers/residual.py]: No reason

File: models/src/anemoi/models/layers/sparse_projector.py (hunk 7 above)
Comment (JPXKQX) [reply models/src/anemoi/models/layers/sparse_projector.py]: No

File: models/src/anemoi/models/layers/residual.py [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+    autocast : bool, default False
+        Whether to use automatic mixed precision for the projections.
+    truncation_up_file_path : str, optional
+        File path (.npz) to load the up-projection matrix from.
+    truncation_down_file_path : str, optional
+        File path (.npz) to load the down-projection matrix from.
+
+    Example
+    -------
+    >>> from torch_geometric.data import HeteroData
```
Comment (anaprietonem) [inline models/src/anemoi/models/layers/residual.py:88]: Very nice docstring!

File: models/src/anemoi/models/layers/residual.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+class NoConnection(BaseResidualConnection):
+    """No-op connection
+
+    This module returns a zero tensor with the same shape as the last timestep.
+    """
+
+    def forward(self, x: torch.Tensor, grid_shard_shapes=None, model_comm_group=None) -> torch.Tensor:
+        """Return a zero tensor with the same shape as the last timestep."""
+        return torch.zeros_like(x[:, 0, ...], device=x.device, dtype=x.dtype)
```
Comment (anaprietonem) [inline models/src/anemoi/models/layers/residual.py]: Not too critical but in the future it could be nice to think about how to define the NoConnection more efficiently and then avoid instantiating a big array in the GPU that's not used

File: models/src/anemoi/models/layers/residual.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        return up_edges, down_edges
+
+    def forward(self, x: torch.Tensor, grid_shard_shapes=None, model_comm_group=None) -> torch.Tensor:
+        """Apply truncated skip connection."""
+        batch_size = x.shape[0]
+        x = x[:, -1, ...]  # pick latest step
+        shard_shapes = apply_shard_shapes(x, 0, grid_shard_shapes) if grid_shard_shapes is not None else None
+
+        x = einops.rearrange(x, "batch ensemble grid features -> (batch ensemble) grid features")
+        x = self._to_channel_shards(x, shard_shapes, model_comm_group)
```
Comment (anaprietonem) [inline models/src/anemoi/models/layers/residual.py:189]: what's the difference between `_to_grid_shards`  and `_to_channel_shards`  - as in I understand once would sharding across the grid dimension and the other across the num channels - but if the shard_shapes is the same between the two calls - then ? maybe @japols can chime in

File: models/src/anemoi/models/layers/residual.py (hunk 10 above)
Comment (anaprietonem) [reply models/src/anemoi/models/layers/residual.py]: or just adding 0 maybe?
````
import torch
a=torch.ones(3,3)
a=a.to('cuda')
assert a==a+0
```
seems fine

Comment (JPXKQX) [reply models/src/anemoi/models/layers/residual.py]: The current code needs the residual connection output to have multiple dimensions. For example,

```
        x_out[..., self._internal_output_idx] += x_skip[..., self._internal_input_idx]
```

Comment (japols) [reply models/src/anemoi/models/layers/residual.py]: could we just return None and only apply skip connection if not None?

File: models/src/anemoi/models/layers/residual.py (hunk 11 above)
Comment (ssmmnn11) [reply models/src/anemoi/models/layers/residual.py:189]: we go from full channels but sharded grid points to full grid points but sharded channels. shard_shapes only carry information about the geometry.

File: models/src/anemoi/models/layers/residual.py [hunk 12]
```diff
@@ -0,0 +1,215 @@
+# (C) Copyright 2025 Anemoi contributors.
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+from abc import ABC
```
Comment (HCookie) [inline models/src/anemoi/models/layers/residual.py:11]:
```suggestion

from abc import ABC
```

File: models/src/anemoi/models/layers/residual.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+    grid to a truncated (coarse) grid, and another to project back to the original resolution.
+
+    Parameters
+    ----------
+    graph : HeteroData
+        The graph containing the subgraphs for down and up projections.
+    data_nodes : str
+        Name of the nodes representing the data nodes.
+    truncation_nodes : str
+        Name of the nodes representing the truncated (coarse) nodes.
```
Comment (HCookie) [inline models/src/anemoi/models/layers/residual.py]: From the code, these are also optional

File: models/src/anemoi/models/layers/sparse_projector.py [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+    edges_name: Optional[tuple[str, str, str]] = None,
+    edge_weight_attribute: Optional[str] = None,
+    **kwargs,
+) -> SparseProjector:
+    """Factory method to build a SparseProjector."""
+    assert (file_path is not None) ^ (
+        graph is not None and edges_name is not None
+    ), "Either file_path or graph and edges_name must be provided."
+
+    if file_path is not None:
```
Comment (HCookie) [inline models/src/anemoi/models/layers/sparse_projector.py:145]: Arguably this should also check for the setting of the graph in addition to / instead of the callers of this function

File: models/src/anemoi/models/layers/residual.py (hunk 12 above)
Comment (JPXKQX) [reply models/src/anemoi/models/layers/residual.py:11]: ?

Comment (HCookie) [reply models/src/anemoi/models/layers/residual.py:11]: ?

**Review by HCookie — DISMISSED** [review]
